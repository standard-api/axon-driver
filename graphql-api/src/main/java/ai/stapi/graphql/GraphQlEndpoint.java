package ai.stapi.graphql;

import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;


@RestController
public class GraphQlEndpoint {

  private final GraphQlExecutor graphQlExecutor;

  public GraphQlEndpoint(
      GraphQlExecutor graphQlExecutor
  ) {
    this.graphQlExecutor = graphQlExecutor;
  }

  @PostMapping("/graphql")
  @ResponseBody
  public Map<String, Object> graphQl(@RequestBody GraphQlOperation graphQLOperation) {
    var response = new LinkedHashMap<String, Object>();
    response.put("data", this.graphQlExecutor.execute(graphQLOperation));
    return response;
  }
}
