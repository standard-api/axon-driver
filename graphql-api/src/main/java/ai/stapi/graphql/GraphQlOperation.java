package ai.stapi.graphql;

import java.util.HashMap;
import java.util.Map;

public class GraphQlOperation {

  private String operationName;
  private String query;

  private Map<String, Object> variables;

  protected GraphQlOperation() {
  }

  public GraphQlOperation(String operationName, String query, Map<String, Object> variables) {
    this.operationName = operationName;
    this.query = query;
    this.variables = variables;
  }

  public GraphQlOperation(String operationName, String query) {
    this(operationName, query, new HashMap<>());
  }

  public String getQuery() {
    return query;
  }

  public String getOperationName() {
    return operationName;
  }

  public Map<String, Object> getVariables() {
    return variables;
  }
}
