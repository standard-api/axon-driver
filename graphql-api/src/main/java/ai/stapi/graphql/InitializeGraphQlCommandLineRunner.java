package ai.stapi.graphql;

import ai.stapi.graphql.graphqlJava.graphQLProvider.GraphQLProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.CommandLineRunner;

public class InitializeGraphQlCommandLineRunner implements CommandLineRunner {

  private final GraphQLProvider graphQLProvider;

  private final Logger logger;

  public InitializeGraphQlCommandLineRunner(
      GraphQLProvider graphQLProvider
  ) {
    this.graphQLProvider = graphQLProvider;
    this.logger = LoggerFactory.getLogger(InitializeGraphQlCommandLineRunner.class);
  }

  @Override
  public void run(String... args) {
    try {
      this.graphQLProvider.initialize();
    } catch (Exception exception) {
      this.logger.error("unable to generate graphql schema", exception);
    }
  }
}
