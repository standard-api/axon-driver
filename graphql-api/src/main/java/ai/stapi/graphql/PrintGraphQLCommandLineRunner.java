package ai.stapi.graphql;

import ai.stapi.graphql.generateGraphQlSchema.PrintGraphQlSchema;
import java.util.Arrays;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ApplicationContext;


public class PrintGraphQLCommandLineRunner implements CommandLineRunner {

  private final CommandGateway commandGateway;

  private final ApplicationContext applicationContext;

  public PrintGraphQLCommandLineRunner(
      CommandGateway commandGateway,
      ApplicationContext applicationContext
  ) {
    this.commandGateway = commandGateway;
    this.applicationContext = applicationContext;
  }

  @Override
  public void run(String... args) {
    if (args.length == 0) {
      throw new RuntimeException("Please specify path where to print schema as command line argument.");
    }
    var path = Arrays.stream(args)
        .filter(arg -> arg.startsWith("_schemaOutputPath:"))
        .map(arg -> arg.replace("_schemaOutputPath:", ""))
        .findFirst();

    var command = new PrintGraphQlSchema(path.orElse(args[0]));
    this.commandGateway.sendAndWait(command);
    var exitCode = SpringApplication.exit(this.applicationContext, () -> 0);
    System.exit(exitCode);
  }
}
