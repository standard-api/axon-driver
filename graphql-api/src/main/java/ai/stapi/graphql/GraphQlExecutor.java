package ai.stapi.graphql;

import ai.stapi.graphql.graphqlJava.graphQLProvider.GraphQLProvider;
import graphql.ExecutionInput;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class GraphQlExecutor {

  private final GraphQLProvider schemaProvider;

  public GraphQlExecutor(GraphQLProvider schemaProvider) {
    this.schemaProvider = schemaProvider;
  }

  public Map<Object, Object> execute(GraphQlOperation graphQLOperation) {
    var executionInputBuilder = new ExecutionInput.Builder().query(graphQLOperation.getQuery());

    if (graphQLOperation.getOperationName() != null) {
      executionInputBuilder.operationName(graphQLOperation.getOperationName());
    }
    if (graphQLOperation.getVariables() != null) {
      executionInputBuilder.variables(graphQLOperation.getVariables());
    }
    var executionResult = this.schemaProvider.getGraphQL().execute(executionInputBuilder);
    return executionResult.getData();
  }

}
