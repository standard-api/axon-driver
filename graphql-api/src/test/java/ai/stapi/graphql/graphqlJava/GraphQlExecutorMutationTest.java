package ai.stapi.graphql.graphqlJava;

import ai.stapi.graphql.GraphQlExecutor;
import ai.stapi.graphql.GraphQlOperation;
import ai.stapi.graphql.graphqlJava.testfixtures.TestGraphqlModelDefinitionsLoader;
import ai.stapi.test.domain.DomainTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(TestGraphqlModelDefinitionsLoader.SCOPE)
class GraphQlExecutorMutationTest extends DomainTestCase {

  @Autowired
  private GraphQlExecutor graphQlExecutor;

  @Test
  void itCanExecuteCreationalMutation() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                mutation {
                  createOrganization(
                    id: "pragueMedicalClinic",
                    payload: {
                      address: [
                        {
                          line: [
                            "Vaclavske Namesti 8",
                            "Vaclavske Namesti 11",
                            "Vaclavske Namesti 50"
                          ],
                          city: "Prague",
                          state: "Praha",
                          postalCode: "11000",
                          country: "Czech Republic",
                          position: {
                            latitude: 50.08804,
                            longitude: 14.42076
                          }
                        }
                      ],
                      telecom: [
                        {
                          system: "email",
                          value: "info@praguemedicalclinic.cz"
                        }
                      ]
                    }
                  ) {
                    successes
                  }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenLastEventGraphApproved();
  }
  
  @Test
  void itCanExecuteCreationalMutationForStructureDefinition() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                mutation {
                  createStructureDefinition(
                    id: "MyEntity",
                    payload: {
                      abstract: false,
                      baseDefinition: "http://hl7.org/fhir/StructureDefinition/DomainResource",
                      baseDefinitionRef: "DomainResource",
                      description: "An example entity",
                      differential: {
                        element: [
                          {
                            definition: "This is long description of an example entity (in most common language), or aggregate (in domain-driven design language), or resource (in HL7 FHIR standard). It will appear in graphQL documentation.",
                            id: "MyEntity",
                            isModifier: false,
                            max: "*",
                            min: 0,
                            mustSupport: false,
                            path: "MyEntity",
                            short: "his is long description of MyEntity"
                          },
                          {
                            definition: "A name associated with the MyEntity.",
                            id: "MyEntity.name",
                            isModifier: false,
                            isSummary: true,
                            max: "1",
                            min: 0,
                            mustSupport: false,
                            path: "MyEntity.name",
                            requirements: "Need to use the name as the label of the MyEntity.",
                            short: "Name used for the MyEntity resource",
                            type: [
                              {
                                code: "string",
                                codeRef: "string"
                              }
                            ]
                          }
                        ]
                      },
                      experimental: true,
                      kind: "resource",
                      name: "MyEntity",
                      status: "draft",
                      type: "MyEntity",
                      url: "http://myorganization.org/fhir/StructureDefinition/MyEntity",
                      version: "0.0.1",
                    }
                  ) {
                    successes
                  }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenLastEventGraphApproved();
  }

  @Test
  @Disabled
  void itCanExecuteCreationalMutationForStructureDefinitionWithDate() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                mutation {
                  createStructureDefinition(
                    id: "MyEntity",
                    payload: {
                      abstract: false,
                      baseDefinition: "http://hl7.org/fhir/StructureDefinition/DomainResource",
                      baseDefinitionRef: "DomainResource",
                      date: "2023-06-11T12:47:40",
                      description: "An example entity",
                      differential: {
                        element: [
                          {
                            definition: "This is long description of an example entity (in most common language), or aggregate (in domain-driven design language), or resource (in HL7 FHIR standard). It will appear in graphQL documentation.",
                            id: "MyEntity",
                            isModifier: false,
                            max: "*",
                            min: 0,
                            mustSupport: false,
                            path: "MyEntity",
                            short: "his is long description of MyEntity"
                          },
                          {
                            definition: "A name associated with the MyEntity.",
                            id: "MyEntity.name",
                            isModifier: false,
                            isSummary: true,
                            max: "1",
                            min: 0,
                            mustSupport: false,
                            path: "MyEntity.name",
                            requirements: "Need to use the name as the label of the MyEntity.",
                            short: "Name used for the MyEntity resource",
                            type: [
                              {
                                code: "string",
                                codeRef: "string"
                              }
                            ]
                          }
                        ]
                      },
                      experimental: true,
                      kind: "resource",
                      name: "MyEntity",
                      status: "draft",
                      type: "MyEntity",
                      url: "http://myorganization.org/fhir/StructureDefinition/MyEntity",
                      version: "0.0.1",
                    }
                  ) {
                    successes
                  }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenLastEventGraphApproved();
  }

  @Test
  void itCanExecuteCreationalMutationWithUnionParameter() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                mutation {
                  createOrganization(
                    id: "pragueMedicalClinic",
                    payload: {
                      level: { string: "VeryMuch" }
                    }
                  ) {
                    successes
                  }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenLastEventGraphApproved();
  }

  @Test
  void itCanExecuteCreationalMutationWithComplexParameterWithUnionField() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                mutation {
                  createOrganization(
                    id: "ExampleOrganization",
                    payload: {
                      address: [
                        {
                          extension: [
                              {
                                  url: "http://urlto.coding.extension",
                                  value: {
                                      Coding: {
                                          display: "DisplayName"
                                      }
                                  }
                              },
                              {
                                  url: "http://urlto.string.extension",
                                  value: {
                                      string: "Brnenice"
                                  }
                              }
                          ]
                          country: "Czech Republic",
                          city: "Brno"
                        }
                      ]
                    }
                  ) {
                    successes
                  }
                }
            """
    );
    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenLastEventGraphApproved();
  }
}