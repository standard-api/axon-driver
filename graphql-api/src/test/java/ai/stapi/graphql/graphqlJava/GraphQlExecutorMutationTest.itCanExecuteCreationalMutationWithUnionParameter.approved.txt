nodes:
	node_type: Organization
	node_hash: 1A39C
	node_edges:
	attributes:
		level -> VeryMuch

edges:
